### PR TITLE
Improve WASI support

### PR DESCRIPTION
**Pull request by electrum**
This implements enough of WASI to run Python 3.13 to _"Hello world"_, including the creation of `.pyc` cached bytecode files.

Given Java's limited file system support, it doesn't seem possible to safely implement sandboxing in the presence of hostile, concurrent access to the local file system. Paths containing symbolic links will also be a problem after `path_symlink()` is implemented. It may be possible to improve this with `SecureDirectoryStream`, but it doesn't support all operations, and is currently only supported for Linux, so I gave up on that for now. [Jimfs](https://github.com/google/jimfs) is a safe option for untrusted code.

I'm putting up this PR now for collaboration and early feedback. We still need tests for this, possibly utilizing [wasi-testsuite](https://github.com/WebAssembly/wasi-testsuite).

### CODE REVIEW

**Comment by andreaTP** [comment]
Thanks for sharing this work @electrum !
Looks super interesting and promising, a couple of considerations:
- yesterday we discussed the fact that the `wasi` module will stay in this GH repository only for an incubating period, and we probably want to take it to a more "community contributed" repo
- testing is missing and the `wasi-testsuite` looks a great starting point, are you going to give a spin at integrating it? Do you need any help in doing so?

**Comment by electrum** [comment]
I'm happy to contribute to this wherever the code lives. Is there an action you want me to take here, or is this just FYI?

I can start working on `wasi-testsuite` integration. It might make sense to extend `test-gen-plugin` to generate test sources. That would be more complicated if this is in a different repo, or could complicate moving the module. Another potential option is to use `@TestFactory` or similar to generate the test methods. Thoughts?

**Comment by andreaTP** [comment]
>  Is there an action you want me to take here, or is this just FYI?

Just FYI :+1:

> It might make sense to extend test-gen-plugin to generate test sources.

Not sure how much you are going to reuse from there, a similar Maven Plugin but targeting directly the `wasi` module is gonna be easier and simpler to extract when time comes. But this is more of a gut feeling than anything ...

I would stick with plain code generation as it demonstrated to work really well with the tooling, but open to alternatives if they bring benefits.

I noticed that the `wasi-testsuite` depends heavily on Python, as soon as we work with a venv(or equivalent) I think that it can be a "mandatory dependency" to avoid complicating things any further.

I plan(hopefully) to spend some time on this project next week, feel free to share any WIP and I'll do my best assisting your effort as we go  😊 it's valuable, thanks a lot!

**Comment by andreaTP** [comment]
@electrum I took a quick but closer look at the `wasi-testsuite` and, probably, that the infrastructure developed for the `fuzz` testing is covering 70-80% of the use-case.

If(please double-check) I'm right it would be nice to avoid the dependency on Python and depend only on the testcases.

Let me know what you think!

**Comment by electrum** [comment]
I agree that we don't need the Python code. Implementing our own test runner should be simple, and they also document how to do that.

**Comment by andreaTP** [comment]
@electrum I'm finally able to spend some cycles on this project after a little pause.
I'm interested in the `wasi-testsuite` integration, if you have not started yet I can give it a shot.

Let me know!

**Review by andreaTP — CHANGES_REQUESTED** [review]
Amazing work @electrum , minor comments but it looks almost finished to me 🎉 
Let's try to draft a quick plan:
- we resolve the final comments

either:
- we merge this effort and we split it out soon
- we create a separate repository and redirect this PR

I don't have strong opinions and I'm happy to move forward and take the additional work when necessary, but we need @bhelx (currently on holiday, back next week I think) support in making it happen.

As an alternative, we can fork and split this repo under a personal account to contain only the wasi stuff and transfer it back to `dylibso` later on.

Proposed repo name: `chicory-wasi-contrib`
wdyt?

**Comment by andreaTP** [comment]
Forgot to mention that the CI is failing on Windows, I understand that supporting it might be challenging.
I'm open to disable the Windows support if it's too much work at this point in time.

File: wasi-test-gen-plugin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      <version>${jackson.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-databind</artifactId>
+      <version>${jackson.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>com.squareup</groupId>
+      <artifactId>javapoet</artifactId>
```
[the remark this answers is not in the document]
Comment (electrum) [reply wasi-test-gen-plugin/pom.xml:34]: I've used Javapoet successfully in other projects and it seems to have a nicer API, automatically handles imports, etc. This usage doesn't require parsing existing Java code, so it seemed like a better fit. But I certainly understand the value of consistency, so I'm not opposed to changing it.

File: wasi-test-gen-plugin/src/main/java/com/dylibso/chicory/maven/WasiTestGenMojo.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.shared.model.fileset.FileSet;
+import org.apache.maven.shared.model.fileset.util.FileSetManager;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.errors.ConfigInvalidException;
+
+/**
+ * This plugin generates test classes for the WASI test suite.
+ */
+@Mojo(name = "wasm-test-gen", defaultPhase = GENERATE_TEST_SOURCES, threadSafe = true)
```
[the remark this answers is not in the document]
Comment (electrum) [reply wasi-test-gen-plugin/src/main/java/com/dylibso/chicory/maven/WasiTestGenMojo.java:41]: Good catch, I missed updating that one.

File: wasi/pom.xml [hunk 3]
```diff
@@ -22,6 +22,11 @@
       <groupId>org.junit.jupiter</groupId>
       <artifactId>junit-jupiter-engine</artifactId>
     </dependency>
+    <dependency>
+      <groupId>com.google.jimfs</groupId>
+      <artifactId>jimfs</artifactId>
+      <version>1.3.0</version>
```
[the remark this answers is not in the document]
Comment (electrum) [reply wasi/pom.xml:28]: Thanks, that should indeed be test scope. I wasn't sure if you wanted `dependencyManagement` for dependencies that are only used in a single module.

File: wasi/src/main/java/com/dylibso/chicory/wasi/Files.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                    @Override
+                    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+                            throws IOException {
+                        if (isSymbolicLink(dir)) {
+                            return FileVisitResult.SKIP_SUBTREE;
+                        }
+
+                        Path directory = target.resolve(source.relativize(dir).toString());
+
+                        if (!directory.toString().equals("/")) {
```
[the remark this answers is not in the document]
Comment (electrum) [reply wasi/src/main/java/com/dylibso/chicory/wasi/Files.java:35]: Good point, I'm not sure. Though I don't think any of this code will work if run against a Windows file system as it requires reading `unix:*` attributes. It should run fine on Windows if using Jimfs for the sandbox file system, which is what we should recommend anyway, as the APIs cannot be implemented in a concurrency-safe way, due to limitations in the Java file system APIs.

**Comment by electrum** [comment]
I'll try to fix the CI failing on Windows. Since we're using Jimfs, there shouldn't be any dependencies on the native behavior.

Regarding merging now vs splitting out the repo, I'm fine either way. I do have reservations about naming this `-contrib`. That feels like it's "this is unsupported, use at your own risk". If the goal of the Chicory project is to have a usable WASM environment for the JVM, then WASI support is import.

Given that preview1 is stable and basically frozen, what are the concerns with supporting it? What is the motivation for splitting it out into a separate project?

File: wasi-test-gen-plugin/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply wasi-test-gen-plugin/pom.xml:34]: I don't have specific issues with Javapoet, but I'm using JavaParser in a bunch of projects and it boils down to consistency as you mentioned, opened https://github.com/electrum/chicory/pull/2 to spare you this work.

**Comment by electrum** [comment]
I'm happy to join the meeting if I can make it.

Let me try to expand my feelings around the name "contrib". It reminds me of things like [zlib contrib](https://github.com/madler/zlib/tree/develop/contrib), which is a collection of random things that users have contributed to the project over the years. They're often years out of date, won't compile, untested, unreviewed, unmaintained, etc.

When I say "unsupported", I mean that from the perspective of the open source project. It says to me, _"here's some random code that we can't say anything about"_. This is not the case for the WASI implementation, which follows the same practices as the rest of the project. It's reviewed, tested, maintained, etc. We want people to use it, will help them where possible, etc.

This doesn't mean that the WASI implementation needs to be 100% complete, pass all the tests, or do everything that someone might want. It can have limitations, possibly documented, like the rest of Chicory or any other project.

Does the Chicory project want to ship and maintain a WASI implementation?

File: wasi-test-gen-plugin/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (electrum) [reply wasi-test-gen-plugin/pom.xml:34]: Thanks, I'll squash that here

**Comment by electrum** [comment]
This should be ready now. I fixed the CI for Windows.

**Comment by andreaTP** [comment]
> They're often years out of date, won't compile, untested, unreviewed, unmaintained, etc.

Ok, I was certainly not thinking about such a bad scenario, thanks for sharing the context.
Maybe, having clear docs about the current status and support is enough to settle expectations.

**Review by andreaTP — APPROVED** [review]
Amazing job, LGTM! 🎉

**Comment by andreaTP** [comment]
Thanks a lot for this great contribution @electrum 🚀

**Comment by bhelx** [comment]
Just catching up as I'm back from vacation. This is such excellent work thank you so much for contributing! Excited to try some of this out.

My general feeling is it can stay in the repo for the time being. I don't feel super strongly about it but I think most people expect things like random, clocks, std io, etc to work out of the box (esp if they are going to use something like python). So having some well supported implementation of WASIp1 easy to include is a nice thing. My primary concern with WASI support is security and correctness. But I think those things can be mitigated with bright red lines in the readmes about our level of confidence about the different parts of the codebase. I would consider moving it if it grows to be very big or we want to focus more on a future version of WASI.

We can discuss more in our meeting today, i have a lot to catch up on. @electrum feel free to drop by if you want. I'll post a link to the zoom in the Zulip. We are meeting at 17:00 UTC today.

**Comment by electrum** [comment]
I missed seeing the meeting invite until just now. Hope I can attend next week.

The implementation should be secure as long as it is used with Jimfs or another sandboxed file system. All file system access is done relative to the `Path` instances passed as preopened directories.

**Comment by bhelx** [comment]
> I missed seeing the meeting invite until just now. Hope I can attend next week.

No problem! Send me your email in Zulip if you want me to add you to invite list.

> The implementation should be secure as long as it is used with Jimfs or another sandboxed file system. All file system access is done relative to the Path instances passed as preopened directories.

Yes, I think that's a welcome dependency that I hadn't thought to use! I think this decision alone already makes it more secure than say how node.js supports Wasi. Certainly alleviates a lot of concerns about the filesytem calls. My general concern is more around supporting any system imports at the moment as this is the typical vector for abuse in these types of sandboxed systems. We haven't put any resources into auditing or verifying the security, but I think that just calling this out in the README should be enough for this stage of the project. I'm going to be writing up some readme changes this week and I'll ping you on the changes related to wasi support for your input.

**Pull request merged**